<p1>To use MLP, CNN, RNN, RNN with LSTM</p1>

In [26]:
import pandas as pd
import numpy as np
from sklearn import linear_model
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split 
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import seaborn as sn
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectPercentile, chi2
import tensorflow as tf

In [29]:
#Function to preprocess the data
def PreProc(X, Y):
    scaler = StandardScaler().fit(X)
    Columns_Scaled = scaler.transform(X)
    #pca = PCA(n_components = 1000).fit(Columns_Scaled).transform(Columns_Scaled)
    #Columns_Scaled = pca
    return Columns_Scaled

<h1>RNN with LSTM</h1>

In [30]:
df = pd.read_csv("emotions.csv")
#columns will be the features
columns = df.drop("label", axis = 1)
print(df.shape)
print(columns.shape)
#y will be the labels, which will be changed to 0,1,2 to represent the negative, neutral and positive state respectively
y = df.label
y = y.replace(["NEGATIVE", "NEUTRAL", "POSITIVE"], [0,1,2])
#Pre-processing columns
columns = PreProc(columns, y)
#splitting the data for training and testing purposes
X_Train, X_Test, Y_Train, Y_Test = train_test_split(columns, y, test_size = 0.2)
#converting to tensors
X = tf.constant(X_Train)
Y = tf.constant(Y_Train)
print(X.shape)
print(Y.shape)
#reshaping X tensor to fit the (batch_size, timestep, feature_length) format
X = tf.reshape(X, (X.shape[0], 1, X.shape[1]))
#adding a LSTM layer and a Dense Layer
model = tf.keras.Sequential()
model.add(tf.keras.layers.LSTM(units=5, input_shape=(1, X.shape[2])))
model.add(tf.keras.layers.Dense(units = 3, activation = 'softmax'))
model.compile(loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X, Y, epochs=100)
X_Test = tf.reshape(X_Test, (X_Test.shape[0], 1, X_Test.shape[1]))
y_pred = model.predict(X_Test)
check = []
for i in y_pred:
    check.append(np.where(i==max(i))[0][0])
print(classification_report(Y_Test, check))

(2132, 2549)
(2132, 2548)
(1705, 2548)
(1705,)


C:\Users\kaust\AppData\Local\Temp\ipykernel_25964\1888380280.py:8: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  y = y.replace(["NEGATIVE", "NEUTRAL", "POSITIVE"], [0,1,2])
C:\Users\kaust\AppData\Local\Programs\Python\Python310\lib\site-packages\keras\src\layers\rnn\rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7797 - loss: 0.7257   
Epoch 2/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9324 - loss: 0.4261 
Epoch 3/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9460 - loss: 0.3649 
Epoch 4/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9570 - loss: 0.3000 
Epoch 5/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9588 - loss: 0.2593 
Epoch 6/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9568 - loss: 0.2319 
Epoch 7/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9533 - loss: 0.2180 
Epoch 8/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9674 - loss: 0.1780 
Epoch 9/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9736 - loss: 0.1515 
Epoch 10/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9709 - loss: 0.1425 
Epoch 11/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9830 - loss: 0.1198 
Epoch 12/100
54/54 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/ste

In [ ]:
model = tf.keras.Sequential()
model.add(tf.keras.layers.LSTM(units=5, input_shape=(1, 1000)))
#model.add(tf.keras.layers.Dense(units = 5, activation = 'softmax'))
model.compile(loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X, Y, epochs=100)

<h1>Conclusion</h1>
<p1>Training accuracy according to study : 97.27%</p1>
<br></br>
<p1>Testing accuracy according to study : 87%</p1>
<br></br>
<p1>Training accuracy obtained : 99.61%</p1>
<br></br>
<p1>Testing accuracy obtained : 98%</p1>
<br></br>
<p1>N.B.: The model used here used both a LSTM layer and a Dense layer. When used without the dense layer, the accuracy reaches approx. 92%. It can be inferred that the addition of a Dense layer greatly increased the accuracy obtained</p1>

<h1>RNN</h1>

<h1>CNN</h1>

<h1>MLP</h1>